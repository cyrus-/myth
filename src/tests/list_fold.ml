let s = "
type nat =
  | O
  | S of nat

type bool =
  | True
  | False

type list =
  | Nil
  | Cons of nat * list

let rec sum (n1:nat) (n2:nat) : nat =
  match n1 with
  | O -> n2
  | S (n1) -> S (sum n1 n2)
;;

let rec is_odd (n:nat) : bool =
  match n with
  | O -> False
  | S (n) ->
    (match n with
    | O -> True
    | S (n1) -> is_odd n1)
;;

let count_odd : nat -> nat -> nat =
  fun (n1:nat) -> fun (n2:nat) ->
    match is_odd n2 with
    | True -> S (n1)
    | False -> n1
;;

let list_fold : (nat -> nat -> nat) -> nat -> list -> nat |>
    { sum => ( 0 => ( [] => 0
                    | [1] => 1
                    | [2; 1] => 3
                    | [3; 2; 1] => 6 )
             | 1 => [] => 1 )
    | count_odd => ( 0 => ( [] => 0
                          | [1] => 1
                          | [2; 1] => 1
                          | [3; 2; 1] => 2 ) )
    } = ?"
