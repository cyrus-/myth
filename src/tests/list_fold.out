(* Generated Thu Nov 13 00:35:27 2014 *)
let list_fold : (nat -> nat -> nat) -> nat -> list -> nat =
  fun (f2:nat -> nat -> nat) ->
    fun (n1:nat) ->
      let rec f4 (l1:list) : nat =
        match l1 with
          | Nil -> n1
          | Cons (n2, l2) -> f2 (f4 l2) n2
      in
        f4
;;
